rice-datanode/source/data_transfer_server.cc

```diff
@@ -318,7 +322,121 @@ void TransferServer::synchronize(std::function<void(TransferServer&, tcp::socket
+    base_proto->set_allocated_block(&blockToTarget);
+    ::hadoop::common::TokenProto *token_header = base_proto->mutable_token();
+
+    token_header->set_identifier("open");
+    token_header->set_password("sesame");
+    token_header->set_kind("foo");
+    token_header->set_service("bar");
+
+    // send the read request
+    uint16_t version = 1000; // TODO what is the version
+    unsigned char read_request = READ_BLOCK;
+    read_block_proto.SerializeToString(&read_string);
+    LOG(INFO) << " writing read request " << std::to_string(read_request);
+    if (!write_header(sock, version, read_request)) {
+        LOG(ERROR) << " could not write the header request to target datanode";
+        return false;
+    } else {
+        LOG(INFO) << " successfully wrote the header request to target datanode";
+    }
+
+    if (!rpcserver::write_delimited_proto(sock, read_string)) {
+        LOG(ERROR) << " could not write the delimited read request to target datanode";
+        return false;
+    } else {
+        LOG(INFO) << " successfully wrote the delimited read request to target datanode";
+    }
+
+    // read the read response
+    BlockOpResponseProto read_response;
+    if (!rpcserver::read_delimited_proto(sock, read_response)) { 
+        LOG(ERROR) << " could not read the read response from target datanode";
+        return false;
+    } else {
+        LOG(INFO) << " successfully read the read response from target datanode";
+    }
+    std::string data(len, 0);
+
+    // read in the packets while we still have them, and we haven't processed the final packet
+    int read_len = 0;
+    bool last_packet = false;
+    while (read_len < len && !last_packet) {
+        asio::error_code error;
+        uint32_t payload_len;
+        rpcserver::read_int32(sock, &payload_len);
+        uint16_t header_len;
+        rpcserver::read_int16(sock, &header_len);
+        PacketHeaderProto p_head;
+        rpcserver::read_proto(sock, p_head, header_len);
+        // LOG(INFO) << "Receiving packet " << p_head.seqno();
+        // read in the data
+        if (!p_head.lastpacketinblock()) {
+            uint64_t data_len = p_head.datalen();
+            uint64_t seqno = p_head.seqno();
+            uint64_t offset = p_head.offsetinblock();
+            error = rpcserver::read_full(sock, asio::buffer(&data[offset], data_len));
+            if (error) {
+                LOG(ERROR) << "Failed to read packet " << p_head.seqno();
+                return false;
+            }
+            read_len += data_len;
+        } else {
+            last_packet = true;
+        }
+    }
+
+    //TODO send ClientReadStatusProto (delimited)
+    if (!fs->writeBlock(header->baseheader().block().blockid(), data)) {
+        LOG(ERROR) << "Failed to allocate block " << header->baseheader().block().blockid();
+    } else {
+        dn->blockReceived(header->baseheader().block().blockid(), read_len);
+    }
+
+    base_proto->release_block();
```

Comment:
if you use set_allocated_block then protobuf thinks you malloc'd it and will try to free on destruction. but it's not a malloc'd memory so tell it we've already released it.
refer to commit message on https://github.com/Rice-Comp413-2016/RDFS/commit/2297a700ad76456958ce851e148ad01f6811fdd2